File: AppleXcode/KOmegaBFEx.playground/Sources/kw.swift
import Foundation
import vsip
func oddNumber(_ n: Int) -> Bool {
    return (n % 2) == 1
}
public class KW {
    public let Nfreq: Int                /* Nts/2 + 1              */
    public let Navg:Int                  /* Scale factor for number of averages */
    public var cm_freq: OpaquePointer?   /* (Nsens, Nfreq) col maj vsip_cmview_d */
    public var rm_freq: OpaquePointer?   /* (Nsens, Nts)   rwo maj vsip_mview_d */
    public var m_gram: OpaquePointer?    /* (Nsens, Nfreq) col maj */
    public var rcfftm: OpaquePointer?    /* by row Nsens by Nts    */
    var ccfftm: OpaquePointer?    /* by col Nsens by Nfreq  */
    public var ts_taper: OpaquePointer?  /* of length Nts          */
    public var array_taper: OpaquePointer? /* of length Nsens        */
    public init(param: Param){
        if oddNumber(param.Nts!){
            preconditionFailure("Data Length must be even")
        }
        Navg = param.Navg!
        Nfreq = param.Nts!/2 + 1
        if let cm_freq = vsip_cmcreate_d(vsip_length(param.Nsens!),vsip_length(Nfreq),VSIP_COL,VSIP_MEM_NONE) {
            self.cm_freq = cm_freq
            if let rm_freq = vsip_mrealview_d(self.cm_freq){
                self.rm_freq = rm_freq
            }
        }
        if let m_gram = vsip_mcreate_d(vsip_length(param.Nsens!), vsip_length(Nfreq), VSIP_COL,VSIP_MEM_NONE){
            self.m_gram = m_gram
        }
        if let rcfftm = vsip_rcfftmop_create_d(vsip_length(param.Nsens!),vsip_length(param.Nts!),1,VSIP_ROW,0,vsip_alg_hint(rawValue: 0)) {
            self.rcfftm = rcfftm
        }
        if let ccfftm = vsip_ccfftmip_create_d(vsip_length(param.Nsens!),vsip_length(Nfreq),1.0,VSIP_FFT_FWD,VSIP_COL,0,vsip_alg_hint(rawValue: 0)){
            self.ccfftm = ccfftm
        }
        if let ts_taper   = vsip_vcreate_hanning_d(vsip_length(param.Nts!),VSIP_MEM_NONE){
            self.ts_taper = ts_taper
        }
        if let array_taper = vsip_vcreate_hanning_d(vsip_length(param.Nsens!),VSIP_MEM_NONE) {
            self.array_taper = array_taper
        }
    }
    deinit {
        vsip_malldestroy_d(m_gram);
        vsip_fftm_destroy_d(rcfftm);
        vsip_fftm_destroy_d(ccfftm);
        vsip_valldestroy_d(ts_taper);
        vsip_valldestroy_d(array_taper);
        vsip_mdestroy_d(rm_freq);
        vsip_cmalldestroy_d(cm_freq);
    }
    public func komega(m_data: OpaquePointer){
        /* Data tapers for time and space */
        vsip_vmmul_d(self.ts_taper,m_data,VSIP_ROW,m_data)
        vsip_vmmul_d(self.array_taper,m_data,VSIP_COL,m_data)
        
        /* FFT for time and space */
        vsip_rcfftmop_d(rcfftm,m_data,cm_freq);
        vsip_ccfftmip_d(ccfftm,cm_freq)
        
        /* power estimate */
        /* to save memory place estimate in real part of spectral matrix */
        vsip_mcmagsq_d(cm_freq,rm_freq)
        
        /* scaling for average */
        vsip_smmul_d(1.0/Double(Navg),rm_freq,rm_freq)
        /* add in new values to gram estimate */
        vsip_madd_d(rm_freq,m_gram,m_gram)
    }
    public func zero(){
         vsip_mfill_d(0.0,m_gram)
    }
    public func instance() -> OpaquePointer {
        return self.m_gram!
    }
}

File: AppleXcode/KOmegaBFEx.playground/Resources/kw.c
#include"kw.h"

void kw_zero(kw_obj *kw){
   vsip_mfill_f(0.0,kw->m_gram);
}

void komega( kw_obj* kw, vsip_mview_f *m_data){
    /* Data tapers for time and space */
    vsip_vmmul_f(kw->ts_taper,m_data,VSIP_ROW,m_data);
    vsip_vmmul_f(kw->array_taper,m_data,VSIP_COL,m_data);

    /* FFT for time and space */
    vsip_rcfftmop_f(kw->rcfftm,m_data,kw->cm_freq);
    vsip_ccfftmip_f(kw->ccfftm,kw->cm_freq);

    /* power estimate */
    /* to save memory place estimate in real part of spectral matrix */
    vsip_mcmagsq_f(kw->cm_freq,kw->rm_freq);

    /* scaling for average */
    vsip_smmul_f(1.0/(float)kw->Navg,kw->rm_freq,kw->rm_freq);
    /* add in new values to gram estimate */
    vsip_madd_f(kw->rm_freq,kw->m_gram,kw->m_gram);
    return;
}

/* initialize the komega object */
int kw_init(kw_obj* kw, param_obj* param){
     int retval = 0;
     vsip_length Nsens = param->Nsens;
     vsip_length Nts   = param->Nts;
     vsip_length Nfreq;
     vsip_length L = Nsens * Nfreq;
     if(param->Nts % 2){
          printf("Data length must be even\n"); 
          retval++;
     }
     kw->Nfreq = Nts/2 + 1;
     kw->Navg = param->Navg;
     Nfreq = kw->Nfreq;
     kw->cm_freq = vsip_cmcreate_f(Nsens,Nfreq,VSIP_COL,VSIP_MEM_NONE);
     if(kw->cm_freq){
         kw->rm_freq=vsip_mrealview_f(kw->cm_freq);
         if(!kw->rm_freq) retval++;
     } else {
         retval++;
         kw->rm_freq = NULL;
     }
     kw->m_gram = vsip_mcreate_f(Nsens, Nfreq, VSIP_COL,VSIP_MEM_NONE);
     if(!kw->m_gram) retval++;
     kw->rcfftm = vsip_rcfftmop_create_f(Nsens,Nts,1,VSIP_ROW,0,0);
     if(!kw->rcfftm) retval++;
     kw->ccfftm = vsip_ccfftmip_create_f(Nsens,Nfreq,1.0,VSIP_FFT_FWD,VSIP_COL,0,0);
     if(!kw->ccfftm) retval++;
     kw->ts_taper   = vsip_vcreate_hanning_f(Nts,0);
     if(!kw->ts_taper) retval++;
     kw->array_taper = vsip_vcreate_hanning_f(Nsens,0);
     if(!kw->array_taper) retval++;
     if(retval) printf("Error in kw initialization\n");
     return retval;
}

/* finalize the komega object */
void kw_fin( kw_obj *kw){
     vsip_malldestroy_f(kw->m_gram);
     vsip_fftm_destroy_f(kw->rcfftm);
     vsip_fftm_destroy_f(kw->ccfftm);
     vsip_valldestroy_f(kw->ts_taper);
     vsip_valldestroy_f(kw->array_taper);
     vsip_mdestroy_f(kw->rm_freq);
     vsip_cmalldestroy_f(kw->cm_freq);
     return;
}

vsip_mview_f *kw_instance(kw_obj* kw){
    return kw->m_gram;
};
